prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1408 @@
+
+The extra request "keep packet header length stable" (stable header length) was introduced in libfabric 1.13.0
+release, and was assigned the ID 2.
+
+This extra request would be useful if endpoint want to implement the "zero copy receive" optimization.
+
+As can be seen from previous discussion, because EFA device does not support ordered delivery, an endpoint
+usually needs to use a temporary buffer to receive incoming packets, and copy data to application's receive buffer
+later. However, if an application has the following set of requirement:
+
+   1. does not need ordered send/receive (`FI_ORDER_SAS`).
+   2. only send/receive eager messages.
+   3. does not use tagged send.
+   4. does not require FI_DIRECTED_RECV (the ability to receive only from certain address).
+
+it should be possible to receive data directly using application buffer directly. Because under such condition,
+receiver does not have special requirement on the data it is going to receive, thus will accept any message from
+the sender.
+
+However, there is one more hurdle to overcome in implementing "zero copy receive", which is the packet header
+length.
+
+Under the condition that endpoints "will only send eager messages" and "do not use tag matching", a sender will
+send data in an EAGER_MSGRTM packet. An EAGER_MSGRTM packet is consisted of packet header and application data.
+However, we cannot put packet header in application's receive buffer. Therefore, for "zero copy receive" to work,
+the receiver need to:
+
+   a. be able to estimate the packer header length of an incoming EAGER_MSGRTM packet, and the
+   b. the packet header length of EAGER_MSGRTM cannot change throughout the communication.
+
+However, there is no guarantee in the base protocol that the packet header length of EAGER_MSGRTM will not
+change.
+
+In fact, because of the existance of the handshake sub-protocol, the packet header length of an EAGER_MSGRTM
+will definitely change. Recall handshake sub-protocol's workflow is:
+
+Before receiving handshake packet, an endpoint will always include the optional raw address header in REQ packets.
+
+After receiving handshake packet, an endpoint will stop including the optional raw address header in REQ packets.
+
+The extra feature "keep packet header length stable" (stable header length) is designed to solve this problem.
+
+When an endpoint toggle on this extra request, its peer will try to satisfy it by keep the header length stable.
+Exactly how to achieve that is up to the implementation to decide, the easiest way to do that is to keep including
+raw address header in the EAGER_MSGRTM even after receiving the handshake packet.
+
+Note, because this is an extra request, an endpoint cannot assume its peer will comply with the request. Therefore,
+the receiving endpoint must be able to handle the situation that a received packet does not have the expected header length.
```

Comment:
what does "stable header length" mean? Given a certain protocol (e.g. v4) it seems there should be a known and constant header length then -- not just "stable"

Reply:
You are right. "stable" is not the correct word here. I changed to constant in new revision.